src/lib/operators/join_helper/join_output_writing.cpp

```diff
@@ -0,0 +1,308 @@
+
+#include <unordered_map>
+
+#include <boost/functional/hash_fwd.hpp>
+
+#include "hyrise.hpp"
+#include "scheduler/job_task.hpp"
+#include "storage/segment_iterate.hpp"
+
+namespace {
+
+using namespace hyrise;  // NOLINT(build/namespaces)
+
+using PosLists = std::vector<std::shared_ptr<const AbstractPosList>>;
+using PosListsByColumn = std::vector<std::shared_ptr<PosLists>>;
+
+struct PosListsHasher {
+  size_t operator()(const PosLists& pos_lists) const {
+    return boost::hash_range(pos_lists.begin(), pos_lists.end());
+  }
+};
+
+/**
+ * Returns a vector where each entry with index i references a PosLists object. If entries i and j reference the same
+ * PosLists object, both columns originate from the same table. As a consequence, they can later share the same (newly
+ * created) PosLists of the join result.
+ * @param input_table
+ */
+PosListsByColumn setup_pos_list_mapping(const std::shared_ptr<const Table>& input_table) {
+  Assert(input_table->type() == TableType::References, "Function only works for reference tables");
+
+  const auto chunk_count = input_table->chunk_count();
+  const auto column_count = input_table->column_count();
+
+  std::unordered_map<PosLists, std::shared_ptr<PosLists>, PosListsHasher> shared_pos_lists_by_pos_lists;
+  shared_pos_lists_by_pos_lists.reserve(column_count);
+
+  auto pos_lists_by_column = PosListsByColumn(column_count);
+  auto pos_lists_by_column_it = pos_lists_by_column.begin();
+
+  // For every column, for every chunk
+  for (auto column_id = ColumnID{0}; column_id < column_count; ++column_id) {
+    // Get all the input pos lists so that we only have to pointer cast the segments once. Without storing the
+    // pointers, we would need to fetch the position lists by fetching them from the input table (get chunk,
+    // get segment, cast segment to obtain PosList) in write_output_segments.
+    auto pos_list_ptrs = std::make_shared<PosLists>();
+    pos_list_ptrs->reserve(chunk_count);
+
+    // Iterate over every chunk and add the chunks segment with column_id to pos_list_ptrs
+    for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+      const auto chunk = input_table->get_chunk(chunk_id);
+      Assert(chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
+
+      const auto& ref_segment_uncasted = chunk->get_segment(column_id);
+      const auto ref_segment = std::static_pointer_cast<const ReferenceSegment>(ref_segment_uncasted);
+      pos_list_ptrs->push_back(ref_segment->pos_list());
+    }
+
+    // pos_list_ptrs contains all position lists of the reference segments for the current column at `column_id`.
+    auto iter = shared_pos_lists_by_pos_lists.emplace(*pos_list_ptrs, pos_list_ptrs).first;
+
+    *pos_lists_by_column_it = iter->second;
+    ++pos_lists_by_column_it;
+  }
+
+  return pos_lists_by_column;
+}
+
+/**
+ *
+ * @param output_segments [in/out] Vector to which the newly created reference segments will be written.
```

Comment:
Pick here, found across whole file: sometimes there are full stops, sometimes there are none.

Reply:
Appended them to full sentences. Did not use them for doxygen stuff (I think we do not have a clear rule for that :-/).

Reply:
Was just a random pick here, found stops/no stops in other comments